### DISCUSSION TITLE
Is there an alternative Golang middleware to continuous REST calls?

### DISCUSSION THREAD

**Discussion by frederikhors** [ory/kratos#1310] [Q&A] [answered]
_I apologize if I open this issue but I can't find the answer on the docs._

Let's say I have Kratos running on my system and people can register. And this is the auth-kratos "microservice".

Then I have my monolith written in Golang which uses middleware before handlers (routes) to check if a user is logged-in or not.

QUESTIONS:

- Do I necessarily have to use REST `/whoami` calls to find out if it is authenticated?

- Can I have a middleware inside my Golang application that analyzes the cookie and tells if it is OK or not?

- If this is not possible, is an HTTP-2/3 gRPC handler provided in order to have a much faster response from the microservice?

I'm worrying about speed because this check will happen for every call to my application.

**Comment by zepatrik** [comment]
You do have to check with Kratos, as that is the only way to reliably log out users. The cookie itself is opaque so there is no way to check it in your code. You can use e.g. Oathkeeper to encode the user information in a JWT instead to use in a micro service architecture, but that would not be helpful in your setting with a monolith.
gRPC APIs would be nice to have though, and as we recently used them in another project, Keto, we would like to add them here as well. If there is no issue for that yet, can you maybe open one?

**Reply by frederikhors** [reply]
What is the purpose of the client like this for Go? https://github.com/ory/kratos-client-go

**Reply by aeneasr** [reply]
It's the official SDK for Ory Kratos

**Reply by frederikhors** [reply]
It is a real regret that Kratos is written in Go and cannot be used in a simple Golang application, intended for example for a standalone distribution on various devices.

**Reply by frederikhors** [reply]
Is there a way?

**Reply by aeneasr** [reply]
>It is a real regret that Kratos is written in Go and cannot be used in a simple Golang application, intended for example for a standalone distribution on various devices.

That's an assumption based on limited exposure to the project :) Of course this is possible. And, to be honest, I don't understand what you mean that it is not possible.

> Do I necessarily have to use REST /whoami calls to find out if it is authenticated?

Was answered by: You do have to check with Kratos, as that is the only way to reliably log out users. The cookie itself is opaque so there is no way to check it in your code.

> Can I have a middleware inside my Golang application that analyzes the cookie and tells if it is OK or not?

Yes you can, using for example the kratos-client-go SDK or by reading the **quickstart** where we explain this as the first thing: https://www.ory.sh/kratos/docs/quickstart#setup

> If this is not possible, is an HTTP-2/3 gRPC handler provided in order to have a much faster response from the microservice?

We currently do not support gRPC but might in the future if there is evidence that the effort to add gRPC is worth it (right now there is no such evidence).

**Reply by frederikhors** [reply]
> That's an assumption based on limited exposure to the project :) Of course this is possible. And, to be honest, I don't understand what you mean that it is not possible.

In all the docs pages I only find ways to install Kratos as a "sidecar" to my application: running it on a dedicated port, waiting for `whoami` calls.

What I'm looking for is a way to bundle my app in a **single executable**.

Is there a way to integrate Kratos this way?

This is the question I cannot find answer.

**Reply by aeneasr** [reply]
I see - you want to embed the Ory Kratos binary in your Go program? While theoretically possible (we actually do this ourselves for the SaaS) we highly discourage you from doing that as we have constant breaking changes in the internal APIs. Meaning, that it will be a lot of work for you to keep Ory Kratos updated due to that.

It's also generally a good idea to keep Ory Kratos running as a single process so that it is an isolated, given that it handles your user's PII and passwords.

So yes, it's possible, but not documented by intent - because we do not officially support it and we discourage people from doing this as we have seen too many people fail with this approach.

**Reply by frederikhors** [reply]
> So yes, it's possible, but not documented by intent - because we do not officially support it and we discourage people from doing this as we have seen too many people fail with this approach.

This is the regret I was talking about earlier.

For small installations on small devices I think it is a good solution to embed Kratos.

Can I ask you to reconsider? And to prepare this possibility for the future?

**Reply by aeneasr** [reply]
As I said above, this is already possible but we simply can not make any guarantees to internal API stability. Maintaining REST API stability is in itself difficult enough - not being able to use e.g. Go 1.17 or Go 2.0 because other people rely on using Ory Kratos as an embedded Go program is not practicable.

If you want to do this anyways and accept the downsides - difficult to update, lots of work to integrate - then it is certainly possible. I just don't recommend it!

Ory Kratos already supports UNIX sockets, so if you run on a small device (e.g. Raspberry PI) you can run kratos as a separate process, bind it to the unix port, and be done with it :) Versus tons of work to integrate, maintain, and update!

**Reply by frederikhors** [reply]
Ok. But two lines on how to do it would not be bad.

For example, I could create a dedicated microservice that connects to my main one via a persistent gRPC connection so as to be much faster than now.

Can we add these tiny instructions on the docs?
I really don't know where to start if I should do it now.

**Comment by aeneasr** [comment, marked as answer]
Answered by:

- https://github.com/ory/kratos/discussions/1310#discussioncomment-693738
- https://github.com/ory/kratos/discussions/1310#discussioncomment-696910
